### PR TITLE
fix: improve-date-histo-request

### PR DESCRIPTION
**Pull request by GGui**
close gravitee-io/issues#2138

### CODE REVIEW

File: gravitee-repository-elasticsearch/src/main/java/io/gravitee/repository/elasticsearch/analytics/query/DateHistogramQueryCommand.java [hunk 1]
```diff
@@ -47,7 +47,14 @@
 	@Override
 	public DateHistogramResponse executeQuery(Query<DateHistogramResponse> query) throws AnalyticsException {
 		final DateHistogramQuery dateHistogramQuery = (DateHistogramQuery) query;
-		final String sQuery = this.createQuery(TEMPLATE, query);
+s
```
Comment (aelamrani) [inline gravitee-repository-elasticsearch/src/main/java/io/gravitee/repository/elasticsearch/analytics/query/DateHistogramQueryCommand.java]: A failed cmd + s?

File: gravitee-repository-elasticsearch/src/main/java/io/gravitee/repository/elasticsearch/healthcheck/query/AbstractElasticsearchQueryCommand.java [hunk 2]
```diff
@@ -77,4 +77,16 @@ protected String createQuery(final String templateName, final Query<T> query) {
 		
 		return request;
 	}
+
+	String createQuery(final String templateName, final Query<T> query, Long newFrom, Long newTo ) {
```
Comment (aelamrani) [inline gravitee-repository-elasticsearch/src/main/java/io/gravitee/repository/elasticsearch/healthcheck/query/AbstractElasticsearchQueryCommand.java]: Should be in ../analytics/query/AbstractElasticsearchQueryCommand.java

File: gravitee-repository-elasticsearch/src/main/java/io/gravitee/repository/elasticsearch/analytics/query/DateHistogramQueryCommand.java [hunk 3]
```diff
@@ -47,7 +47,14 @@
...
+		final Long from = dateHistogramQuery.timeRange().range().from();
+		final Long to = dateHistogramQuery.timeRange().range().to();
+
+		final long newFrom = (from.longValue() / dateHistogramQuery.timeRange().interval().toMillis()) * dateHistogramQuery.timeRange().interval().toMillis();
```
Comment (aelamrani) [inline gravitee-repository-elasticsearch/src/main/java/io/gravitee/repository/elasticsearch/analytics/query/DateHistogramQueryCommand.java]: newFrom/newTo => roundedFrom/roundedTo?

File: gravitee-repository-elasticsearch/src/main/java/io/gravitee/repository/elasticsearch/analytics/query/DateHistogramQueryCommand.java [hunk 4]
```diff
@@ -47,7 +47,14 @@
 	@Override
 	public DateHistogramResponse executeQuery(Query<DateHistogramResponse> query) throws AnalyticsException {
 		final DateHistogramQuery dateHistogramQuery = (DateHistogramQuery) query;
-		final String sQuery = this.createQuery(TEMPLATE, query);
+s
+		final Long from = dateHistogramQuery.timeRange().range().from();
+		final Long to = dateHistogramQuery.timeRange().range().to();
+
```
Comment (aelamrani) [inline gravitee-repository-elasticsearch/src/main/java/io/gravitee/repository/elasticsearch/analytics/query/DateHistogramQueryCommand.java]: Add a small comment to explain the benefits of this, by using internal ES cache...

File: gravitee-repository-elasticsearch/src/main/java/io/gravitee/repository/elasticsearch/analytics/query/AbstractElasticsearchQueryCommand.java [hunk 5]
```diff
@@ -82,12 +82,27 @@
...
+
+	/**
+	 * Create the elasticsearch query
+	 * @param templateName Freemarker template name
+	 * @param query query parameter
+	 * @param newFrom from parameter
+	 * @param newTo to parameter
+	 * @return the elasticsearch json query
+	 */
+	String createQuery(final String templateName, final Query<T> query, Long newFrom, Long newTo ) {
```
Comment (aelamrani) [inline gravitee-repository-elasticsearch/src/main/java/io/gravitee/repository/elasticsearch/analytics/query/AbstractElasticsearchQueryCommand.java]: newFrom => roundedFrom

File: gravitee-repository-elasticsearch/src/main/java/io/gravitee/repository/elasticsearch/healthcheck/query/AbstractElasticsearchQueryCommand.java [hunk 6]
```diff
@@ -77,4 +77,16 @@ protected String createQuery(final String templateName, final Query<T> query) {
...
+		final Map<String, Object> data = new HashMap<>();
+		data.put("query", query);
+		data.put("newFrom", newFrom);
```
Comment (aelamrani) [inline gravitee-repository-elasticsearch/src/main/java/io/gravitee/repository/elasticsearch/healthcheck/query/AbstractElasticsearchQueryCommand.java]: newFrom => roundedFrom

File: gravitee-repository-elasticsearch/src/main/resources/freemarker/dateHistogram.ftl [hunk 7]
```diff
@@ -20,8 +20,8 @@
         {
           "range": {
             "@timestamp": {
-              "from": ${query.timeRange().range().from()},
-              "to": ${query.timeRange().range().to()},
+              "from": ${newFrom},
```
Comment (aelamrani) [inline gravitee-repository-elasticsearch/src/main/resources/freemarker/dateHistogram.ftl]: newFrom/newTo => roundedFrom/roundedTo

**Comment by aelamrani** [comment]
What about having this capabilities also for the HC?

File: gravitee-repository-elasticsearch/src/main/java/io/gravitee/repository/elasticsearch/healthcheck/query/AbstractElasticsearchQueryCommand.java (hunk 2 above)
Comment (aelamrani) [reply gravitee-repository-elasticsearch/src/main/java/io/gravitee/repository/elasticsearch/healthcheck/query/AbstractElasticsearchQueryCommand.java]: when i made the review that did not compile... you changed createQuery method in ../**analytics**/query/DateHistogramQueryCommand.java while you change the signature in ../**healthcheck**/query/AbstractElasticsearchQueryCommand.java (and not in ./**analytics**/query/AbstractElasticsearchQueryCommand.java)

**Comment by GGui** [comment]
what is "HC"?

**Comment by aelamrani** [comment]
Sorry, i meant Health Check. What about having the same optimization for the HC search?

**Pull request merged**